### PR TITLE
Replace visited instructions with ancestry sets

### LINKED ISSUE #34: Replace visited instructions with ancestry sets
The implementation of cycle detection using the visited instructions inside Trace is computational and memory heavy. This can be optimized using a DisjointSet data structure.

### PR DESCRIPTION
**Pull request by Martinn1996**
This fixed #34.

This is a implementation to detect cycles in a graph by using a Union-Find algorithm with a Disjoint Set. 
Also, the test coverage is now 100%.

### CODE REVIEW

File: src/Generation.ts [hunk 1]
```diff
@@ -108,23 +109,41 @@ export default class Generation {
 		// If a thread for pc already exists in this generation, combine traces and return
 		const existingThreadForProgramCounter = this._threadsByProgramCounter[pc];
 		if (existingThreadForProgramCounter) {
-			// Detect repetition in the same generation, which would cause cyclic traces
-			if (!parentThread || !parentThread.trace.contains(pc, this._generationNumber)) {
-				// Non-cyclic trace, join threads
-				existingThreadForProgramCounter.join(parentThread, badness);
+			if (parentThread === null || parentThread === undefined) {
```
Comment (DrRataplan) [inline src/Generation.ts]: This can stay as a `!parentThread` check imo

File: test/DisjointSet.tests.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        it('Rank array is initialized', () => {
+            let disjointSet = new DisjointSet(7);
+            expect(disjointSet.rank.length).toBe(7);
+            expect(disjointSet.rank).toEqual([0, 0, 0, 0, 0, 0, 0]);
+        });
+    });
+
+    describe('Union and Find', () => {
+        it('Element is the parent itself', () => {
+            let disjointSet = new DisjointSet(7);
```
Comment (DrRataplan) [inline test/DisjointSet.tests.ts]: These may be consts

File: test/Examples.tests.ts [hunk 3]
```diff
@@ -329,7 +329,7 @@ describe('whynot.js examples', () => {
 			// returns any traces, these represent how the program was able to match the input. If
 			// it doesn't, the input did not match in any way.
 			const vm = compileRegexVM('(a|b)*', false);
-
+
```
Comment (DrRataplan) [inline test/Examples.tests.ts]: trailing whitespace

File: src/Generation.ts [hunk 4]
```diff
@@ -108,23 +109,41 @@ export default class Generation {
...
+			}
+
+			if (parentThread.generationNumber !== this._generationNumber) {
+				existingThreadForProgramCounter.updateBadness(badness);
+				existingThreadForProgramCounter.trace.join(parentThread.trace);
+				return null;
 			}
 
+			const parentThreadRep = this._ancestries.find(parentThread.threadId);
+			if (parentThreadRep === existingThreadForProgramCounter.threadId) {
```
Comment (DrRataplan) [inline src/Generation.ts]: This here will not correctly detect cycles with the following form:
```
Threads:
1 -> 2
2 -> 3
3 -> 2
```

We can not correctly detect these cycles. If we `find` both threads, we will be able to detect these two threads, but then we will also (incorrectly) detect the following case as a cycle:

```
Thread list:

1 -> 2, 3
2 -> 3
```

Comment (bwrrp) [reply src/Generation.ts]: Good catch! Let's close this PR for now, and revisit / reinvestigate this approach later if / when further optimization is required.

**Comment by bwrrp** [comment]
While promising, it seems that this approach unfortunately has some correctness issues. I'll close this for now until we have time to revisit this optimization.

**Pull request closed without merging**